src/sandbox/functionsSimulations/index.ts

```diff
@@ -0,0 +1,143 @@
+
+import { DEFAULT_PORT } from "../../shared/constants";
+
+const writeSimulationConfigFile = async (
+  hre: HardhatRuntimeEnvironment
+): Promise<string> => {
+  const simulationConfig = hre.config.chainlink.functions_simulation;
+  let tsCodeSimulationsConfig = "";
+  tsCodeSimulationsConfig += `export const secrets = ${JSON.stringify(
+    simulationConfig.secrets || {}
+  )};`;
+  tsCodeSimulationsConfig += `export const maxOnChainResponseBytes = ${
+    simulationConfig.max_on_chain_response_bytes ||
+    functionsToolkit.DEFAULT_MAX_ON_CHAIN_RESPONSE_BYTES
+  };`;
+  tsCodeSimulationsConfig += `export const maxExecutionTimeMs = ${
+    simulationConfig.max_execution_time_ms ||
+    functionsToolkit.DEFAULT_MAX_EXECUTION_DURATION_MS
+  };`;
+  tsCodeSimulationsConfig += `export const maxMemoryUsageMb = ${
+    simulationConfig.max_memory_usage_mb ||
+    functionsToolkit.DEFAULT_MAX_MEMORY_USAGE_MB
+  };`;
+  tsCodeSimulationsConfig += `export const numAllowedQueries = ${
+    simulationConfig.num_allowed_queries ||
+    functionsToolkit.DEFAULT_MAX_HTTP_REQUESTS
+  };`;
+  tsCodeSimulationsConfig += `export const maxQueryDurationMs = ${
+    simulationConfig.max_query_duration_ms ||
+    functionsToolkit.DEFAULT_MAX_HTTP_REQUEST_DURATION_MS
+  };`;
+  tsCodeSimulationsConfig += `export const maxQueryUrlLength = ${
+    simulationConfig.max_query_url_length ||
+    functionsToolkit.DEFAULT_MAX_HTTP_REQUEST_URL_LENGTH
+  };`;
+  tsCodeSimulationsConfig += `export const maxQueryRequestBytes = ${
+    simulationConfig.max_query_request_bytes ||
+    functionsToolkit.DEFAULT_MAX_HTTP_REQUEST_BYTES
+  };`;
+  tsCodeSimulationsConfig += `export const maxQueryResponseBytes = ${
+    simulationConfig.max_query_response_bytes ||
+    functionsToolkit.DEFAULT_MAX_ON_CHAIN_RESPONSE_BYTES
+  };`;
+  const pathToFile = join(__dirname, "/simulationConfig.ts");
+  await fs.promises.writeFile(pathToFile, tsCodeSimulationsConfig);
+  return pathToFile;
+};
+
+export const simulateRequest = async (
+  hre: HardhatRuntimeEnvironment,
+  source: string,
+  args?: string[],
+  bytesArgs?: string[]
+): Promise<DecodedResult> => {
+  const pathToConfigFile = await writeSimulationConfigFile(hre);
+  const simulationDeployment =
+    await functionsToolkit.startLocalFunctionsTestnet(
+      pathToConfigFile,
+      {},
+      hre.config.chainlink.functions_simulation?.port || DEFAULT_PORT
+    );
+
+  const provider = new hre.ethers.providers.JsonRpcProvider(
+    `http://localhost:${
+      hre.config.chainlink.functions_simulation?.port || DEFAULT_PORT
+    }/`
+  );
+  const admin = new hre.ethers.Wallet(
+    simulationDeployment.adminWallet.privateKey,
+    provider
+  );
+  const functionsConsumerAddress =
+    await hre.chainlink.sandbox.functionsConsumer.deploy(
+      simulationDeployment.functionsRouterContract.address,
+      simulationDeployment.donId,
+      {
+        signer: admin,
+        provider,
+      }
+    );
+
+  const functionsConsumer =
```

Comment:
As per previous comment -- the consumer should be whatever the user has written as the Functions consumer? it should be user-provided.

Reply:
Yes, it would be great, I have to think about how to make it in a proper way. As I mentioned earlier, I think that it is necessary to introduce an interface so we will be able to call `sendRequest` method

Reply:
On reflection the only interface strictly required by functions to work is the `fulfillRequest()` callback.  That is the only one that *must* follow the signature/interface spec.  sendRequest can be anything at all as long as the user constructs a Req object correctly.  Technically all the request information can be stored on chain as storage variables, which means there is no interface for the request side, other than creating the `Req` object on chain.

We can specify an interface of *all* the required data for a request, but many of them can be optional(eg: args, secrets, bytesargs etc)